pkg/apis/serving/v1alpha1/inference_graph.go

```diff
@@ -0,0 +1,253 @@
+//
+// Tree Ensemble constitutes a case where simple algorithms for combining results of either classification or regression trees are well known.
+// Multiple classification trees, for example, are commonly combined using a "majority-vote" method.
+// Multiple regression trees are often combined using various averaging techniques.
+// e.g tagging models with segment identifiers and weights to be used for their combination in these ways.
+// ```yaml
+// kind: InferenceGraph
+// metadata:
+//   name: ensemble
+// spec:
+//   nodes:
+//     transformer:
+//       routerType: Sequence
+//       routes:
+//       - service: feast
+//       - nodeName: ensembleModel
+//         data: $response
+//     ensembleModel:
+//       routerType: Ensemble
+//       routes:
+//       - service: sklearn-model
+//       - service: xgboost-model
+// ```
+//
+// Scoring a case using a sequence, or chain, of models allows the output of one model to be passed in as input to subsequent models.
+// ```yaml
+// kind: InferenceGraph
+// metadata:
+//   name: model-chainer
+// spec:
+//   nodes:
+//     mymodel-s1:
+//       routerType: Sequence
+//       routes:
+//       - service: mymodel-s1
+//       - service: mymodel-s2
+//         data: $response
+//       - service: mymodel-s3
+//         data: $response
+// ```
+//
+// In the flow described below, the pre_processing node base64 encodes the image and passes it to two model nodes in the flow.
+// The encoded data is available to both these nodes for classification. The second node i.e. dog-breed-classification takes the
+// original input from the pre_processing node along-with the response from the cat-dog-classification node to do further classification
+// of the dog breed if required.
+// ```yaml
+// kind: InferenceGraph
+// metadata:
+//   name: dog-breed-classification
+// spec:
+//   nodes:
+//     top-level-classifier:
+//       routerType: Sequence
+//       routes:
+//       - service: cat-dog-classifier
+//       - nodeName: breed-classifier
+//         data: $response
+//     breed-classifier:
+//       routerType: Switch
+//       routes:
+//       - service: dog-breed-classifier
+//         condition: { .predictions.class == "dog" }
+//       - service: cat-breed-classifier
+//         condition: { .predictions.class == "cat" }
+// ```
+type InferenceRouter struct {
+	// RouterType
+	//
+	// - `Sequence:` chain multiple inference steps with input/output from previous step
+	//
+	// - `Splitter:` randomly routes to the named service according to the weight
+	//
+	// - `Ensemble:` routes the request to multiple models and then merge the responses
+	//
+	// - `Switch:` routes the request to two or more services with specified routing rule
+	//
+	RouterType InferenceRouterType `json:"routerType"`
+
+	// Routes defines destinations for the current router node
+	// +optional
+	Routes []InferenceStep `json:"routes,omitempty"`
```

Comment:
Should we rename this now?
```suggestion
	Steps []InferenceStep `json:"steps,omitempty"`
```